package/sbp_rtcm3_bridge/src/rtcm3_decode.c

```diff
@@ -0,0 +1,830 @@
+ * \param len Length of bit field in bits.
+ * \param data Signed integer to be packed into bit field.
+ */
+void setbits(u8 *buff, u32 pos, u32 len, s32 data)
+{
+    setbitu(buff, pos, len, (u32)data);
+}
+
+/** Set bit field in buffer from a signed integer.
+ * Packs `len` bits into bit position `pos` from the start of the buffer.
+ * Maximum bit field length is 32 bits, i.e. `len <= 32`.
+ *
+ * \param buff
+ * \param pos Position in buffer of start of bit field in bits.
+ * \param len Length of bit field in bits.
+ * \param data Signed integer to be packed into bit field.
+ */
+void setbitsl(u8 *buff, u32 pos, u32 len, s64 data)
+{
+    setbitul(buff, pos, len, (u64)data);
+}
+
+/** Write RTCM header for observation message types 1001..1004.
+ *
+ * The data message header will be written starting from byte zero of the
+ * buffer. If the buffer also contains a frame header then be sure to pass a
+ * pointer to the start of the data message rather than a pointer to the start
+ * of the frame buffer. The RTCM observation header is 8 bytes (64 bits) long.
+ *
+ * If the Synchronous GNSS Message Flag is set to `0`, it means that no further
+ * GNSS observables referenced to the same Epoch Time will be transmitted. This
+ * enables the receiver to begin processing the data immediately after decoding
+ * the message. If it is set to `1`, it means that the next message will
+ * contain observables of another GNSS source referenced to the same Epoch
+ * Time.
+ *
+ * Divergence-free Smoothing Indicator values:
+ *
+ * Indicator | Meaning
+ * --------- | ----------------------------------
+ *     0     | Divergence-free smoothing not used
+ *     1     | Divergence-free smoothing used
+ *
+ * GPS Smoothing Interval indicator values are listed in RTCM 10403.1 Table
+ * 3.4-4, reproduced here:
+ *
+ * Indicator | Smoothing Interval
+ * --------- | ------------------
+ *  000 (0)  |   No smoothing
+ *  001 (1)  |   < 30 s
+ *  010 (2)  |   30-60 s
+ *  011 (3)  |   1-2 min
+ *  100 (4)  |   2-4 min
+ *  101 (5)  |   4-8 min
+ *  110 (6)  |   >8 min
+ *  111 (7)  |   Unlimited
+ *
+ * \param buff A pointer to the RTCM data message buffer.
+ * \param type Message type number, i.e. 1001..1004 (DF002).
+ * \param id Reference station ID (DF003).
+ * \param t GPS time of epoch (DF004).
+ * \param sync Synchronous GNSS Flag (DF005).
+ * \param n_sat Number of GPS satellites included in the message (DF006).
+ * \param div_free GPS Divergence-free Smoothing Indicator (DF007).
+ * \param smooth GPS Smoothing Interval indicator (DF008).
+ */
+void rtcm3_write_header( const rtcm_msg_header *header, u8 num_sats, u8 *buff )
+{
+  setbitu(buff, 0, 12, header->msg_num);
+  setbitu(buff, 12, 12, header->stn_id);
+  setbitu(buff, 24, 30, (u32)round(header->tow*1e3));
+  setbitu(buff, 54, 1, header->sync);
+  setbitu(buff, 55, 5, num_sats);
+  setbitu(buff, 60, 1, header->div_free);
+  setbitu(buff, 61, 3, header->smooth);
```

Comment:
@pgrgich instead of calculating up these offsets by hand, why not use a value that you pass in that gets updated? e.g., 

```c
u16 bit = 0;
setbitu(buff, &bit, 12, header->msg_num);
setbitu(buff, &bit, 12, header->stn_id);
```

Then `setbitu` would keep the offset updated. Also possible to skip the offset and just updated `buff`. That's what I've done in the past around reading out fields.